smartsim/entity/ensemble.py

```diff
@@ -275,6 +277,29 @@ def attach_generator_files(
                 to_copy=to_copy, to_symlink=to_symlink, to_configure=to_configure
             )
 
+    def print_attached_files(self) -> None:
```

Comment:
Thoughts on adding another method that simply formats the str and then have this method be solely responsible for printing it? Something like:
```py
def attached_files_table(self) -> str:
    # essentially what you have for this method
    # but just return the `tabulate` str
    ...

def print_attached_files(self) -> None:
    print(self.attached_files_table())
```

The only reason I ask is that the former might more closely mimic something like `pandas.DataFrame.head()` in an IPython/Jupyter environment.

Reply:
100% agree w/Matt on splitting finding the attached fileset & printing it but not on example impl.

I'd consider a method that returns the values-of-interest only:
```

def attached files(self) -> t.List[str]:
  per_model_files = {}

  for model in self.models:
    per_model_files[model.name] = []
    if model.files:
        per_model_files[model.name].extend(model.files)

  return per_model_files
```

Now we've got a complete representation of the fileset but it's not commingled with any specific presentation choices. Otherwise, "the ensemble is empty..." makes me have to parse strings to determine if there were files instead of a truthiness check.

Reply:
On second thought, adding the empty message to tabulate may obviate a need for the attached_files method completely since the dict-comprehension is only a single line....

I'd kind of like to build a contract that's an implementation of the visitor pattern and be consistent & clear in where we're offering a tabulating/display API
```
class Tabular(t.Protocol):
  def to_tabular(self) -> Tabular:
    ...
  def on_empty(self) -> str:
    ...

class EnsembleTable(Tabular):
  def to_tabular(self) -> Tabular:
    return {model.name: model.files or [] for model in self.models}
  @property
  def on_empty(self) -> str:
    return "No file attached to this model"

```
Now tabulate would be able to just consume a consistent display API:

```
def tabulate(tabular: Tabular) -> None:
  output = ""
  for thing in tabular:
    output += thing or tabular.empty_msg
```